### PR TITLE
Navigation: add dedicated sidebar for managing layout of navigation menus

### PR DESCRIPTION
**Pull request by gwwar**
## What?

This PR adds a new navigation sidebar in the site editor. This sidebar replaces the list view modal, and allows one to see the contents of navigation menus even if they aren't in the current template.

https://user-images.githubusercontent.com/444434/158396446-53c04c19-5f7b-4748-8f49-1867db28ad14.mov

## Why?

Part of explorations for https://github.com/WordPress/gutenberg/issues/36667.

This PR pulls out a more pragmatic piece of https://github.com/WordPress/gutenberg/pull/38600 out to focus specifically on a Navigation sidebar.

~We may find that it's still a little early to display a new top level panel while we iterate on UX options, so let me know if folks would like to add a feature flag.~ @getdave: I suggest we ship and iterate based on feedback.

## How?

A couple of key points in this PR:
- A new selector `__experimentalGetGlobalBlocksByName` allows us to display the correct current menu when the panel is open.
- Using a `BlockEditorProvider` gives us the editing and save functionality we mostly expect, however **there are side-effects that occur in the rendering of BlockList**. This means that we need to think through how to provide information like allowed block information (which blocks may be inserted into another), when BlockList is not rendered (as is the case for ListView). I've added a temporary workaround by dispatching this information manually. Folks should discuss if this temporary patch is worthwhile, or if we should think of a better solution with some refactoring of block list. Note that the current solution is fragile and will drift, but it does fix dragging and has a relatively small surface area.
- A new slotfill in blocktoolbar to decouple the navigation block from having to know about a site-editor component.
- We only had interest in adding this panel to the site editor, so this PR effectively removes the list view modal in the post-editor without an alternative. @getdave: I think that's ok since we already have the standard List View available in the Post Editor which does the exact same thing.
- ~List View has gained new properties to allow all levels to be collapsed by default.~ (now extracted into https://github.com/WordPress/gutenberg/pull/39486)

## Todo

- [x] Collapse nested levels by default
- [x] Improve "mover" arrows placement for long menu items.
- [ ] ~Disable `Show more settings` option  under Block Actions - doesn't work unless Nav block is selected in editor canvas.~ Out of scope for this PR.
- [x] Extract all changes to `ListView` to a separate PR - now done in:
    -  https://github.com/WordPress/gutenberg/pull/39486.
    - https://github.com/WordPress/gutenberg/pull/39494/
- [x] Extract any other changes that are making this PR overly "noisy" (too many changes):
    - https://github.com/WordPress/gutenberg/pull/39498

## Testing Instructions

All of the below should be tested in the _Site_ Editor.

### Nav block menu to sidebar association

The intent is to verify that the menu matches that associated with the selected Nav block:

- refresh the Site Editor to clear existing Redux state.
- add a Nav block and associate a Navigation Menu with it.
- with the Nav block selected toggle the Nav sidebar open
- see that the menu that is associated with the selected Nav block is active in the sidebar

### Default Navigation Menu selection based on Template

- refresh the Site Editor to clear existing Redux state.
- add a Nav block and associate a Navigation Menu with it.
- _deselect_ the Navigation block.
- toggle the sidebar open.
- the sidebar should select the Nav Menu associated with the first Navigation block in the currently template.
- try variations (i.e. multiple Nav blocks some without associated Menus) to ensure the correct menu is always selected.

### Default Navigation Menu selection based on available Nav Menus

- refresh the Site Editor to clear existing Redux state.
- remove all Nav blocks from the current template.
- toggle the sidebar open.
- see the sidebar Menu displays the first available `wp_navigation` returned by the REST API.

### Modifying Menus items

- Pick a Navigation Menu that you know - temporarily add to a Nav block so you know what it looks like then remove the Nav block.
- Open Nav sidebar.
- _Modify_ the menu - move up and down, delete items...etc
- Check the menu shows up in the multi-entity saving flow.
- Save
- Add the same menu to a new Nav block.
- Check the changes you made in the sidebar are reflected there.

### Misc.

* Verify that in the Widget, Navigation and Post editor that the navigation sidebar is not available.

### CODE REVIEW

**Comment by scruffian** [comment]
Thanks for putting this together

File: packages/block-editor/src/components/block-toolbar/index.js
Comment (getdave) [inline packages/block-editor/src/components/block-toolbar/index.js:155]: This is used to provide a toggle to open the Navigation sidebar for the currently selected Navigation block.

Comment (gwwar) [reply packages/block-editor/src/components/block-toolbar/index.js:155]: Added a picture for folks too. In trunk clicking this opens a list-view modal, but on this branch it should open the new sidebar.

list view

**Review by getdave — COMMENTED** [review]
I took this for a spin. It's a _really_ great start and I can see it provides some benefits as a way to manage a Navigation Menu outside of the Navigation block (which can have its limitations).

Nonetheless there are some rough edges and pain points that we need to either

- fix in this PR
- note down as "known limitations" and track in follow up Issues

This is not intended to sound overly critical. I fully acknowledge that this isn't the "finished article" and we're still iterating on the design and UX of this approach. Moreover, I will likely be working on this PR so it's in my interests to be thorough in exposing problems 😄 

Here are my thoughts:

### UX problems

- the requirement to side scroll to see the ellipsis menu on long menu item names is suboptimal

Screen Capture on 2022-03-10 at 11-06-08

- the loading animation doesn't marry well with the positioning of the menu when it renders

Screen Capture on 2022-03-10 at 11-09-06

- we might want to avoid rendering the menu with all levels fully expanded. This adds a lot of visual noise and makes the menu more difficult to explore for simple operations.
- when toggled open/closed the sidebar seems to have trouble restoring the menu you were last editing 
https://user-images.githubusercontent.com/444434/157650683-436236d6-0678-41b0-90bd-ca026f4d3aeb.mov
- it's possible to toggle block settings from the menu item. When you do this it toggles the block inspector for the block that is currently selected within the _editor_ canvas.
https://user-images.githubusercontent.com/444434/157651417-036f3307-e589-4358-a90d-4f8bd18dd859.mp4
- drag and drop indicators (blue lines) overflow outside of the sidebar
Screen Capture on 2022-03-10 at 11-20-47


### Missing features

These may or may not be within the scope of this PR or indeed the feature more widely. But we should acknowledge them:

- inability to add new menu _items_ - tricky due to inability to edit text...etc
- inability to create new _menus_ - this is important for "safe exploration" (the ability to create menus without having to render them to a post via the Nav block). 


## Summary

I think this is a great start and I appreciate all the effort that's gone into it 👏 

The implementation aside however, even if it were polished I cannot see it solving all the user requirements for managing menus outside of the block. We are missing

- ability to create new menus
- ability to directly manipulate the menu item text content
- ability to manipulate the menu item block settings
- ability to access the raw inner blocks of the Navigation Menu itself

Therefore I wonder whether we should restyle this sidebar less as a means to manage the content of the menu and instead make it more about managing the structure of a menu. Indeed this may have been the original genesis of the idea as outlined in https://github.com/WordPress/gutenberg/issues/36667.

**Comment by gwwar** [comment]
Noting some related issues to ListView that can be worked on outside of this PR:

> we might want to avoid rendering the menu with all levels fully expanded. 

There's some discussion in https://github.com/WordPress/gutenberg/issues/34759 for that. Folks prefer collapsed by default, there will need to be a small refactor for component open/closed state in ListView. Folks should also be mindful of making sure windowing working as expected when ListView has many items.

> the requirement to side scroll to see the ellipsis menu on long menu item names is suboptimal

This is a general issue with ListView taking up an arbitrary width based on the current contents. It's more noticeable here since the right sidebars are smaller than the secondary ones.

I played around with this a bit in https://github.com/WordPress/gutenberg/pull/35605 though I think it's ideal if we can standardize a bit on sidebar widths as well. @jasmussen do you know if there was an existing issue for this one?

> drag and drop indicators (blue lines) overflow outside of the sidebar

We likely only need to fix this if we want this panel to land on the right vs left. Shouldn't be too bad to tidy up.

Also note that folks have been wanting more direct dragging in ListView. I recently updated framer motion to v6 which has new drag utilities, so folks should be able to experiment with that again. Previously performance was a blocker for this, but ListView also uses windowing now for larger cases so that shouldn't be as much of an issue.

Here's a prototype of what it could look like using framer v4: https://github.com/WordPress/gutenberg/pull/34022

>  I wonder whether we should restyle this sidebar less as a means to manage the content of the menu and instead make it more about managing the structure of a menu

🤔 Is that not what the panel currently does?

**Comment by paaljoachim** [comment]
I am testing the PR.

Clicking the "List view" icon in the toolbar opens panel in the sidebar. I would assume that it would be associated with the Navigation block sidebar settings. Clicking the X closes the right sidebar. 

Screenshot 2022-03-10 at 23 03 15

Here it could instead be on top of the panels kinda like clicking the 3 dot contextual Options icon in the top right corner opens on top of the sidebar settings. Here it can remain connected with the Nav block sidebar settings.
Clicking the X would only close the Navigation and one would then see the Nav block sidebar settings underneath.
I removed the Navigation icon seen in the top bar as the space can become too crowded.

Nav-sidebar-panel-on-top-of

Another option.
Could we add the full Navigation as a "regular" panel seen in the standard Navigation block settings? Along with the other Navigation block sidebar settings. Clicking the "List view" icon in the toolbar could then open the panel showing the Navigation. Closing the panel the other Navigation block panels would be seen.

Here is the mockup showing only the one Edit panel.
Edit-nav-in-sidebar

If it can not directly be added to the sidebar panels then one could perhaps have a panel with information and a button to open the Navigation editor.

Nav-block-sidebar-edit-nav

**Comment by scruffian** [comment]
I think the intention of this panel is ultimately that it becomes a "site level" panel as outlined in https://github.com/WordPress/gutenberg/issues/36667. Eventually we'll be able to manage all navigation blocks here, and create new ones. This PR is just a first step towards that goal, and although neither of those things are yet possible we should be looking to move in that direction.

It is easier to get there in small increments, by only changing one thing at at time. I think the next step will probably be to make this panel dark and move it, and the Styles sidebar to the left of the window, so I'd rather keep the behaviour of the sidebar as is for now and revisit it in the next PR.

**Comment by getdave** [comment]
Thank you for the feedback everyone.

I agree we ought to keep the scope of _this_ PR narrow and iterate in future PRs. 

I'm going to look at ironing out bugs and then addressing some of the points that have been raised.

**Comment by getdave** [comment]
Experimented with disabling the `nowrap` on the list view items:
Screen Capture on 2022-03-15 at 11-06-29

```patch
diff --git a/packages/block-editor/src/components/list-view/block-select-button.js b/packages/block-editor/src/components/list-view/block-select-button.js
index 452f01d30b..f4fc1d63d8 100644
--- a/packages/block-editor/src/components/list-view/block-select-button.js
+++ b/packages/block-editor/src/components/list-view/block-select-button.js
@@ -85,7 +85,9 @@ function ListViewBlockSelectButton(
 			>
 				<ListViewExpander onClick={ onToggleExpanded } />
 				<BlockIcon icon={ blockInformation?.icon } showColors />
-				<BlockTitle clientId={ clientId } maximumLength={ 35 } />
+				<span className="block-editor-list-view-block__title">
+					<BlockTitle clientId={ clientId } maximumLength={ 35 } />
+				</span>
 				{ blockInformation?.anchor && (
 					<span className="block-editor-list-view-block-select-button__anchor">
 						{ blockInformation.anchor }
diff --git a/packages/block-editor/src/components/list-view/list-item.js b/packages/block-editor/src/components/list-view/list-item.js
index 6a93b00fd2..185e32a187 100644
--- a/packages/block-editor/src/components/list-view/list-item.js
+++ b/packages/block-editor/src/components/list-view/list-item.js
@@ -42,7 +42,7 @@ export default function ListViewListItem( {
 				onClick={ onClick }
 			>
 				<BlockIcon icon={ blockType?.icon } showColors />
-				{ children ? children : blockLabel }
+				{ children ? children : <span>blockLabel</span> }
 				{ isSelected && (
 					<VisuallyHidden as="span">
 						{ __( '(selected block)' ) }
diff --git a/packages/edit-site/src/components/sidebar/navigation-menu-sidebar/style.scss b/packages/edit-site/src/components/sidebar/navigation-menu-sidebar/style.scss
index 0e227e4f4c..384437b210 100644
--- a/packages/edit-site/src/components/sidebar/navigation-menu-sidebar/style.scss
+++ b/packages/edit-site/src/components/sidebar/navigation-menu-sidebar/style.scss
@@ -26,3 +26,17 @@
 	margin-left: $grid-unit-30;
 	width: 50%;
 }
+
+
+.edit-site-navigation-inspector {
+
+	.block-editor-list-view-leaf .block-editor-list-view-block-contents {
+		align-items: flex-start;
+		white-space: normal;
+	}
+
+	.block-editor-list-view-block__title {
+		margin-top: 3px;
+	}
+}
+
```

**Comment by getdave** [comment]
It would very valuable to gain some design input here. 

Note: we're _specifically_ trying to limit the scope of this PR to encompass adding a place to modify your Navigation Menus _**only**_.

Items where I think design could add most value:

- suggestions for visual improvements to the loading state indicators.
- opinions on positioning of movers and "3 dots" menu.
- general improvements to the aesthetics of the panel (if any).

I will update the PR description with that latest screenshots and information.

cc @jasmussen @jameskoster

**Comment by talldan** [comment]
For a first attempt, I don't think the user experience is too far off. 

I don't recommend trying to ship the block movers or any other new List View features in this PR, maybe a follow-up? It doesn't all have to be done in one go. They've been inactive for probably about two years, I think they'll need some refinement, accessibility testing and e2e tests.

I think it'd be good to consider moving this feature to the left of the screen before merging though. At the moment a user can have two List Views open, which is confusing. I think this one should just replace the global List View when it's selected from the nav block. The designs all show the feature on the left, so I think it's best to start with that so that the ergonomics are correct (e.g. Block Settings can be toggled open). The sidebar on the right is also a bit narrower, so not ideal for this feature.

**Comment by getdave** [comment]
Thanks for taking time to feedback here @talldan.

> I don't recommend trying to ship the block movers or any other new List View features in this PR, maybe a follow-up?

I don't _think_ block movers are a new feature of list view. They come "out of the box" so to speak. What would you suggest? Disabling them?

I could remove other features I've added to ListView into smaller follow ups. That should keep things simple as long as it doesn't muddy the review process. 

I agree that shipping in smaller increments is best.

**Comment by talldan** [comment]
> I don't think block movers are a new feature of list view. They come "out of the box" so to speak. What would you suggest? Disabling them?

Oh right. They've never been shipped as an active feature. I guess they're automatically enabled in the List View component? Maybe they're included here accidentally. I think they can be turned on or off via a `showMovers` prop.

**Comment by getdave** [comment]
Currently this feature is not very exposed to users - let's use that to our advantage. 

We could look to simplify this PR by removing non-essential features (moving those to separate PRs) and shipping "as is".

We can then follow up to:

- re-add non-essential features (based on merging other PRs).
- move the sidebar to the left hand side panel
- design improvements

@talldan @scruffian what do you think?

**Comment by jasmussen** [comment]
Thanks for taking a stab at a minimal first version! Here's what I see in quick GIF form:
state

A few thoughts. Firstly that it seems fine to start the pinned item on the right, even if eventually we'll likely move it left as part of #36667.

Perhaps an artifact of my local env. being sheparded along without a fresh restart in a while, but I can't find in the dropdown list the particular menu used on my site. "Header Navigation" as the chosen menu doesn't show up in the list:

Screenshot 2022-03-16 at 08 12 15

Note that if I choose a menu in the canvas which also shows up on the right, the canvas DOES update to reflect any changes I make dragging things around.

"Navigation" seems like a good simple name for the panel.

The existance of this panel itself seems to suggest a good next step would be to explore: what happens if there are zero navigation menus created saved? Can we show the structure of the site itself (using page metadata such as parent and sort order), and then save a menu as soon as a customization is made?

If I only ever have a single menu saved, we don't need to show the dropdown:

Screenshot 2022-03-16 at 08 16 48

If I select a navigation block in the canvas, and I have multiple menus saved, should the panel update to pick the selected menu?

What do you think? Nice work!

**Comment by talldan** [comment]
> Currently this feature is not very exposed to users - let's use that to our advantage.
> ...
> what do you think?

I think that makes sense. I personally feel this PR could go much smaller, as I'm not sure all the features will make sense to an end user. As an example:
- Selecting a nav block updates the menu in the sidebar, but changing the menu via sidebar's dropdown doesn't equally affect the block
- Using the button on the editor topbar shows the menu for one of the blocks, but doesn't highlight the related nav block.
- You can view a menu that's not in use, but can't do much with that menu.

I think the difficult part is that right now the feature feels very connected to the editor canvas, but the end goal is for it to represent and be used to build a more global site structure. Some of these features are maybe being introduced too early (the menu dropdown being the main one).

**Comment by getdave** [comment]
@jasmussen Thank you for reviewing this 👏 

Worth nothing that in an effort to reduce the complexity of this PR I'm currently cherry picking PRs out of this branch in order to remove some of the features to their own PRs. Once done I will rebase this one which will make it easier to review.

> A few thoughts. Firstly that it seems fine to start the pinned item on the right, even if eventually we'll likely move it left as part of https://github.com/WordPress/gutenberg/issues/36667.

I agree. Start simple and iterate.

> ... I can't find in the dropdown list the particular menu used on my site. "Header Navigation" as the chosen menu doesn't show up in the list:

I haven't experienced that. It requests all the menus from the Navigation endpoint.

> Note that if I choose a menu in the canvas which also shows up on the right, the canvas DOES update to reflect any changes I make dragging things around.

Yes this works - you expected that right?. Any changes made in the sidebar changes the underlying menu data. Therefore anywhere in the editor canvas that uses the same menu data will also update. 

> "Navigation" seems like a good simple name for the panel.

Happy to simplify this. What I will say is that the "W" menu is already identified several times as the "Navigation panel". I'm concerned we're overlapping terminology here which could be especially confusing for those using assistive tech.

> ...what happens if there are zero navigation menus created saved? 

I've updated it to show a message explaining that there are no menus available.

> Can we show the structure of the site itself (using page metadata such as parent and sort order), and then save a menu as soon as a customization is made?

I think this is confusing. This sidebar handles Navigations. If we start showing site structure here then that isn't coupled to a Navigation Menu then I fear it's going to be really confusing when the user adds a Navigation Menu and that whole thing disappears. I still don't fully understand the vision for how this panel is supposed to represent "Site Structure".

> If I only ever have a single menu saved, we don't need to show the dropdown:

Nice catch. Fixed!

> If I select a navigation block in the canvas, and I have multiple menus saved, should the panel update to pick the selected menu?

That should work already. However I think @talldan makes a good point below that this panel seems overly tied to the editor canvas. I don't think we can see this panel as being a fully featured way to manage Navigation _blocks_. Rather it manages to the Navigation Menu data. Therefore we shouldn't couple it to the canvas too much.

**Comment by jasmussen** [comment]
> I haven't experienced that. It requests all the menus from the Navigation endpoint.

A glitch in the matrix! If I find a way to reproduce we can look separately.

> Yes this works - you expected that right?

Absolutely — it was awesome 👌 

> Happy to simplify this. What I will say is that the "W" menu is already identified several times as the "Navigation panel". I'm concerned we're overlapping terminology here which could be especially confusing for those using assistive tech.

Indeed, though I would say the W menu is the problem here, and one that is worth addressing separately as part of #36667. I think there are some opportunities discussed on that thread which might have the W menu be less of a menu that is opened from the site editor, but rather the reverse: the main interface from which the site editor is secondary. This is a bit forward looking, but the main point is that the W menu could use a lot of love.

> I think this is confusing. This sidebar handles Navigations. If we start showing site structure here then that isn't coupled to a Navigation Menu then I fear it's going to be really confusing when the user adds a Navigation Menu and that whole thing disappears. I still don't fully understand the vision for how this panel is supposed to represent "Site Structure".

I'm approaching this from the angle of: what do we show when there's not a single menu defined. When that happens technically in WordPress, the page metadata is used, right? Even in terms of the Page List block being output in that case. In that light it just feels sensible to visualize that. I would acknowledge that you couldn't make a change to that structure, without that action also saving a new menu as part of it. But it'd still provide a starting point that isn't an empty canvas. Does that make sense?

Definitely doesn't have to happen in this PR, but it's something to think about as we need to solve it eventually.

**Comment by getdave** [comment]
Last commit introduced a bug with showing "no menus" message too early. Will fix.

**Comment by getdave** [comment]
I've extracted `__experimentalGetGlobalBlocksByName` to it's own PR https://github.com/WordPress/gutenberg/pull/39610.

This should further simplify this PR.

File: packages/edit-site/src/components/sidebar/navigation-menu-sidebar/style.scss
Comment (scruffian) [inline packages/edit-site/src/components/sidebar/navigation-menu-sidebar/style.scss:1]: We have similar code elsewhere ([packages/block-editor/src/components/link-control/style.scss](https://github.com/WordPress/gutenberg/blob/65f7c31da4b5e0e89ab8b3b6997361834ddd0b00/packages/block-editor/src/components/link-control/style.scss)) so maybe we should make this a mixin?

File: packages/edit-site/src/components/sidebar/navigation-menu-sidebar/navigation-inspector.js
Comment (scruffian) [inline packages/edit-site/src/components/sidebar/navigation-menu-sidebar/navigation-inspector.js:173]: I find all these different cases hard to grok. What do you think about returning early in each case to make it easier to reason about?

**Comment by getdave** [comment]
> I find all these different cases hard to grok. What do you think about returning early in each case to make it easier to reason about?

@scruffian yeh I had the exact same reaction previously. However, you'll see this pattern in various places in Gutenberg for a few reason, principle amongst which is a11y as removing/adding nodes to the DOM can cause focus loss.

An example is [the Template Part block](https://github.com/WordPress/gutenberg/blob/a9b29acd0a058644ef28523c00fbe633272d6c0c/packages/block-library/src/template-part/edit/index.js#L134).

**Review by scruffian — APPROVED** [review]
A few minor comments but this looks great.

**Comment by paaljoachim** [comment]
Retesting.

I really like what I see!

Navigation icon shows up only in the Site Editor screen and not in the posts/pages screen. (I might have been so all the time but it is first now that I noticed the difference.) 
(EDIT: It is good to only see it in the Site Editor. As the posts/pages screen does not need to show it.)

Hovering over the various icons in the top bar. Focus on two or more words capitalization: 
"List View" and "Navigation Menus" but "Toogle block inserter" and "Toggle navigation" does not use capitalization of the second or third word. I believe the focus is on sentence case. Should "Navigation Menus" instead be "Navigation menus"?

The drag and drop in adding a link to the submenu and variations of this as in dragging it out again, or creating a sub sub menu is kind of tricky. As a first version it is ok, but this drag and drop aspect needs to be improved.
Navigation-menus-sidebar

Perhaps the 3 dot contextual drop down could use a method to move a link in/out of the submenu (or creating another level of the submenu). As the the drag and drop method relies on precise movement of the mouse.

I assume we need some accessibility feedback. @alexstine 

Moving a link to a closed submenu: It would be helpful to open the submenu a link is moved into.
Moving a link to an open submenu: Let the submenu remain open.

Good work Dave!

**Comment by jasmussen** [comment]
> Navigation icon shows up only in the Site Editor screen and not in the posts/pages screen.

This is intentional as part of #36667.

**Comment by getdave** [comment]
@jasmussen How are you feeling about this one as a first iteration? Still needs a 👍 from design.

We should list out known follow ups so it's clear that this isn't "finished".

**Comment by jasmussen** [comment]
Looks good!
editor

I wanted to test what happened when there are zero menus saved, and I'm seeing a crash when I go into the advanced section:

crash

This appears to be an issue also in trunk, so technically not a blocker for this one. 

To respond to that state as described, though:

> I've updated it to show a message explaining that there are no menus available.

> > Can we show the structure of the site itself (using page metadata such as parent and sort order), and then save a menu as soon as a customization is made?
> 
> I think this is confusing. This sidebar handles Navigations. If we start showing site structure here then that isn't coupled to a Navigation Menu then I fear it's going to be really confusing when the user adds a Navigation Menu and that whole thing disappears. I still don't fully understand the vision for how this panel is supposed to represent "Site Structure".

It seems totally fine to land this PR with a basic "No menus" message, and keep these ideas as things to potentially explore in the future. It's a great PR, nice work, we should bring about the benefits it comes with!

The opportunity I see here is to look at the trifecta of the Navigation Block, Page metadata, and the site structure inferred from that. In classic menus, if you have zero _menus_ defined, you still see navigation: one that is structured by the "Order" and "Parent" metadata fields from pages:
Screenshot 2022-03-22 at 12 35 21

Some ideas:

* The "Parent" page field is especially useful as when you're manually building out a menu, we can contextually surface the right suggestions when, for example, you're building submenus. Maybe we offer to add all submenus at once, when you just add the parent page.
* We've had some trouble with the empty navigation block in themes and patterns, leading many to preinsert the "All Pages" block in patterns. But even then, you'll often find a need to edit your Header template, select the navigation block, then select the right menu you want to see. 
* Since we're already smart about naming menus ("Header navigation" is saved when you create a new menu from inside the Header template part), we can be smarter with the initial building as well: when you insert a navigation block in a header template part and a "Header navigation" is already saved, default to loading that. Save the user a click: it's easy to change the menu anyway.
* When multiple menus are saved, make a guess! Load the first navigation saved. So long as it's easy to change a menu after the fact, in the case of most simple sites it'll likely guess right!
* We can go a step further in the navigation block. When zero menus are saved, show the Page List by default. Even if in a transient state where clicking the navigation block just shows the existing placeholder we already have (Select menu, Create new, etc). This would bring frontend/editor consistency between what's already happening.

All of that serves to save clicks and interaction. Selecting a menu from the dropdown could become something you do only when we guess wrong the first time. With all that in place, a natural next step feels like surfacing the very same site structure you'd get from a Page List inside the dedicated navigation pane: only, as soon as you make a change, a CPT is saved.

**Comment by getdave** [comment]
@jasmussen Thanks for your 👀  My responses below (and I've taken what you're suggesting into account for follow up work).

**One thing** - are you saying that you'd be happy to give this a 👍 ? If so would you mind doing so? 🙇‍♂️ 

> ...I'm seeing a crash when I go into the advanced section:

As you say, that happens on `trunk`. Need to look into it in another PR.

https://user-images.githubusercontent.com/444434/159503498-ca17067b-4417-4163-84d8-230edfbdf2eb.mp4

> When multiple menus are saved, make a guess! Load the first navigation saved. So long as it's easy to change a menu after the fact, in the case of most simple sites it'll likely guess right!

Just so you know this PR does that already. If Navigation Menus exist it will load the first one unless there is one on the current post/template.

**Comment by jasmussen** [comment]
> (and I've taken what you're suggesting into account for follow up work).

I'd couch it more. An idea for potential followup work. I think they are valid ideas, but there are always obstacles to overcome on the way there!

> One thing - are you saying that you'd be happy to give this a 👍 ? If so would you mind doing so? 🙇‍♂️

Ah yes, I saw an existing green check and omitted the extra approval. One moment.

> Just so you know this PR does that already. If Navigation Menus exist it will load the first one unless there is one on the current post/template.

👌

**Review by jasmussen — APPROVED** [review]

**Comment by getdave** [comment]
# Do not merge

Whilst this is 🟢 I would like to extract the addition of a `<span>` around the block name on the List View (and all the e2e tests changes that were required as a result). 

Why? Because these changes are significant enough that they might need to be rolled back and we should be able to do that via a single focused PR.

The PR is https://github.com/WordPress/gutenberg/pull/39687

**Comment by alexstine** [comment]
Anyone have any suggestions on how I can check accessibility on this? Up to this point, the site editor has been pretty much off limits for most things accessibility related. I have no idea how to even get the block inserted or inserted to what before I can start thinking about new sidebar.

**Comment by getdave** [comment]
**Edit**: I just saw your comment [here](https://github.com/WordPress/gutenberg/issues/39293#issuecomment-1064468572) 😞. I guess that means it's pretty much impossible to review right now without wider improvements. 

> I have no idea how to even get the block inserted or inserted to what before I can start thinking about new sidebar.

Is the Site Editor completely inaccessible or is it just lack of familiarity with it at this stage? Just trying to get a handle on whether this is a fundamental problem or just part of a process whereby you are "learning" the interface.

I'll try and think up a clever way to get a simple setup state in place for you but it might be difficult.

**Comment by tellthemachines** [comment]
> Anyone have any suggestions on how I can check accessibility on this?

@alexstine here are some steps:

1. Activate Twenty Twenty Two theme (it comes with a Navigation block in the header by default, so you don't need to add one);
2. Go to the Site Editor, and tab through the header toolbar until you reach the List View;
3. From there you can find the Navigation block in the header (it's inside a few layers of Group blocks), and press Enter on it to transfer focus to the actual block;
4. Once in the block, switch to Edit mode so you can access the block toolbar; 
5. From the block toolbar, click "Open list view". Currently, that only opens the sidebar with the Navigation-specific list view without transferring focus to it (note that it's not the same list view we were on just previously), but now tabbing twice will take us into that sidebar, which can be navigated like the standard list view.

Another way to reach it should be:

1. Go to Site Editor, and tab through the header toolbar until you reach the "Navigation Menus" toggle. 
2. Currently, on toggling that button, the sidebar opens but focus isn't transferred to it, so that's definitely an a11y issue 😅 

A few observations from my very quick keyboard/visual testing:

* The "Open list view" label on the block toolbar might be confusing because we already have a list view;
* I'd expect focus to be transferred to the sidebar when that button is clicked, as well as from the "Navigation Menus" button in the header;
* When moving inside a navigation item with the Arrow keys, focus doesn't follow the visual order of the buttons. I can only focus "Move up" and "Move down" by focusing the "Options" button first, and then moving back with Left Arrow. Not sure what's happening there, but focus order and markup order aren't matching.
* We might want to think of a shorter text label for the header button, because "Navigation Menus" is taking up a lot of space when button text labels are enabled:
Header toolbar
(we can still keep the full text in the aria-label and show it as a tooltip, just not display the whole thing in the header)

**Comment by getdave** [comment]
@tellthemachines Thank you so much for your review. As this PR is growing ever larger I'm going to look to merge it and then address these points as follow ups. They are important and I will raise Issues for each now and ensure they are addressed as may next tasks.

> The "Open list view" label on the block toolbar might be confusing because we already have a list view;

Ok let's rename that. 

Tracked in: https://github.com/WordPress/gutenberg/issues/39800

> I'd expect focus to be transferred to the sidebar when that button is clicked, as well as from the "Navigation Menus" button in the header

Hmmm. Let me test the other sidebars and see what happens. Should be possible to auto-focus inside the sidebar.

Tracked in: https://github.com/WordPress/gutenberg/issues/39801

> When moving inside a navigation item with the Arrow keys, focus doesn't follow the visual order of the buttons. I can only focus "Move up" and "Move down" by focusing the "Options" button first, and then moving back with Left Arrow. Not sure what's happening there, but focus order and markup order aren't matching.

Is this something we also experience with ListView in other contexts (as in the primary List View)? I'll take a look.

Tracked in https://github.com/WordPress/gutenberg/issues/39802.

> We might want to think of a shorter text label for the header button, because "Navigation Menus" is taking up a lot of space when button text labels are enabled:

Agreed. Great spot.

Tracked in https://github.com/WordPress/gutenberg/issues/39803.

**Comment by tellthemachines** [comment]
Thanks for following up with those issues @getdave !

> Is this something we also experience with ListView in other contexts (as in the primary List View)? I'll take a look.

I'm not sure as the primary List View doesn't have the movers. It occurred to me that this could be related to keyboard improvements in #38358, which changed the behaviour of Left/Right arrows to allow expanding and collapsing rows better.

File: packages/edit-site/src/components/sidebar/navigation-menu-sidebar/navigation-menu.js
Comment (talldan) [inline packages/edit-site/src/components/sidebar/navigation-menu-sidebar/navigation-menu.js:56]: @getdave I think it would be good to disable the block movers until the keyboard navigation issues are solved and there has been accessibility feedback. At the moment the mover buttons don't seem to be in the expected place when using the right arrow key (as mentioned in https://github.com/WordPress/gutenberg/issues/39802).

**Comment by alexstine** [comment]
For accessibility feedback, @tellthemachines covered everything I was going to mention. The whole thing is really not keyboard accessible right now. Is there a tracking issue or new PR yet?

**Comment by getdave** [comment]
> The whole thing is really not keyboard accessible right now. Is there a tracking issue or new PR yet?

Yes I've listed them out [here](
https://github.com/WordPress/gutenberg/pull/39290#issuecomment-1080463373).

**Comment by paaljoachim** [comment]
Switching menus does not affect the current selected Header menu.
I expected that switching a menu in the drop down area when a Navigation block is selected that the menu used in the block would also change.

**Comment by getdave** [comment]
> Switching menus does not affect the current selected Header menu. I expected that switching a menu in the drop down area when a Navigation block is selected that the menu used in the block would also change.

That's not a featured we optimised for. Interesting that you expected that to happen.

To me this whole thing shows that we need to "divorce" the Navigation sidebar from the concept of the editor _canvas_.

**Comment by jasmussen** [comment]
There's a design outlined in https://github.com/WordPress/gutenberg/issues/36667#issuecomment-1087711883 which aims to make it clearer that it is a global interface rather than one contextual to the content like the block inspector is.

**Pull request merged**
